### ISSUE TITLE
Assistance in creating a custom model architecture .

### ISSUE THREAD

**Issue by kailas711** [open-mmlab/mmdetection#10879]
I installed mmdetection after facing 100s of errors and right now for inference i am using a pretrained model to get started with mmdetection. This error keeps on popping up i am passing the argument as str but yet i don't know why it's interpreted as come other dtype 

How to actually work with this library ??

i want to create a custom model more like joining multiple models together, like using their available combination of necks,backbones,detectors etc Is that doable??

![image](https://github.com/open-mmlab/mmdetection/assets/89206677/f6113a13-5eab-44bd-b7d1-d0dbfc5926af)

**Comment by hhaAndroid** [comment]
@kailas711 Are you using the latest dev-3.x branch? This feature is currently available in the dev-3.x branch.

**Comment by kailas711** [comment]
> @kailas711 Are you using the latest dev-3.x branch? This feature is currently available in the dev-3.x branch.

Yes I am sir , I've been looking all over csdn , stackoverflow etc but they're all tutorials of v2x not of v3x and it seems like I am the only one facing this issue

**Comment by Jordan-Pierce** [comment]
Hey @kailas711 is that the demo colab notebook that you're using? If not, would you mind making a sharable link and posting it here. 

For installation, these are what I use:
```python

      - mmcls==0.25.0
      - mmcv==2.0.1
      - mmdet==3.1.0
      - mmengine==0.8.3
      - torch==2.0.1+cu118
      - torchaudio==2.0.2+cu118
      - torchreid==0.2.5
      - torchvision==0.15.2+cu118
```

Regarding inference, here's some of my code, maybe it'll help. It might seem trivial, but make sure to specify the parameters when initiating the inference model (`model=config`) so there is no ambiguity.

```python

import torch
from mmdet.apis import DetInferencer

    try:
        # Find the config file
        config = args.config
        assert os.path.exists(config)
        config_name = os.path.basename(config).split(".")[0]
        print(f"NOTE: Using config file {config_name}")
    except:
        print(f"ERROR: Config file doesn't exist; please check input")
        sys.exit(1)

    try:
        # Find the checkpoint file
        checkpoints = args.checkpoint
        assert os.path.exists(checkpoints)
        checkpoint_name = os.path.basename(checkpoint).split(".")[0]
        print(f"NOTE: Using checkpoint file {checkpoint_name}")
    except:
        print(f"ERROR: Checkpoint file doesn't exist; please check input")
        sys.exit(1)

    # Directory where media is stored
    if os.path.exists(args.media_dir):
        media_dir = args.media_dir
        media_name = os.path.basename(media_dir)
    else:
        print(f"ERROR: Media directory provided doesn't exist; please check input")
        sys.exit(1)

    # Collect the images from frames directory
    if os.path.exists(f"{media_dir}\\frames\\"):
        frame_dir = f"{media_dir}\\frames\\"
        frame_paths = glob.glob(f"{frame_dir}*.*")
        print(f"NOTE: Found {len(frame_paths)} frames in {media_name}")
    else:
        print(f"ERROR: Could not locate 'frames' directory in {media_dir}")
        sys.exit(1)

    # Output directory
    output_dir = args.output_dir
    os.makedirs(output_dir, exist_ok=True)

    # Either GPU or CPU
    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
    print(f"NOTE: Using device {device}")

    try:
        # Load inference object
        inferencer = DetInferencer(model=config,
                                   weights=checkpoint,
                                   palette='coco',
                                   show_progress=True)

        print(f"NOTE: Weights {os.path.basename(checkpoint)} loaded successfully")

    except Exception as e:
        print(f"ERROR: Could not load model\n{e}")
        sys.exit(1)

    try:
        print(f"NOTE: Making predictions for {media_name}")
        # Dict of predictions, and visualization
        results = inferencer(frame_paths,
                             out_dir=output_dir,
                             no_save_pred=False,
                             pred_score_thr=args.pred_threshold)

        print(f"NOTE: Merging predictions")
        merge_predictions(glob.glob(f"{output_dir}preds\\*.json"), class_map, output_dir)

    except Exception as e:
        print(f"ERROR: Could not make prediction on {media_name}\n{e}")
        traceback.print_exc()
```

**Comment by kailas711** [comment]
https://colab.research.google.com/drive/1bJro4-n__sXumyH82B2pO5ar1PAZSsCI?usp=sharing

Here is the link

**Comment by Jordan-Pierce** [comment]
@kailas711 you should be using `mim` to install, not `pip`; see the docs.

Second, you're trying to import Config from mmcv, but if you look at tools/train.py you can see that it's being imported from mmengine

`from mmengine.config import Config, DictAction`

**Comment by kailas711** [comment]
But sir I saw a blog in CSDN( translated to English) 
There they're editing the cofig file , first they change the config/base file and then pass some dictionaries to alter the configuration
Although the blog is using the 2x version of mmdetection
What's shall I do in 3x version

**Comment by kailas711** [comment]
Also sir , is it possible to connect via Google meet or something, i have loads of questions and it would be of immense help sir for my project

**Comment by kailas711** [comment]
@Jordan-Pierce I got the demo notebook running in their in-built/already given dataset (balloons) it uses RTMDet model , i tried changing the model configuration , the backbone from "CSPNeXt" to "ResNeSt" by doing what you suggested. Loading configs and using dict() to define parameters 

![image](https://github.com/open-mmlab/mmdetection/assets/89206677/55f02ed6-67a7-475b-b812-bd9c1fdc09ff)


#### and then i rewrote the config (i believe)

![image](https://github.com/open-mmlab/mmdetection/assets/89206677/091d73ec-f9d3-40df-bafc-00f32407a609)


#### Yet it downloaded the "CSPNext"  instead of "ResNeSt", did i miss something ??? , is it the checkpoint link for resnest??

![image](https://github.com/open-mmlab/mmdetection/assets/89206677/3b354a7a-1e4b-4ba0-aa12-9263cdee1b66)

**Comment by Jordan-Pierce** [comment]
So if you go to the root of the repository you'll find a `configs` folder. In that folder are all the configuration files for all models available. For example, `rtmdet` contains the different versions for the RTMDet models as .py files.

Like in the demo where you loaded the .py file and modified that config file, you'll do the same for any model you want.

Is that what you're doing? I'm sorry, I can't tell based on your screenshots.

**Comment by Jordan-Pierce** [comment]
See here for [ResNest](https://github.com/open-mmlab/mmdetection/blob/main/configs/resnest/README.md). If you look at `config/rtmder/classification/` I believe you're actually just downloading those weights instead of what you're trying to do, which is uses ResNest?

**Comment by Jordan-Pierce** [comment]
See here for [ResNest](https://github.com/open-mmlab/mmdetection/blob/main/configs/resnest/README.md). If you look at `config/rtmder/classification/` I believe you're actually just downloading those weights instead of what you're trying to do, which is uses ResNest? 

CSPNext is just the encoder for RTMDet, and you're writing in the config that you want to use an entirely different architecture (ResNest), but that config file is specifically for RTMDet. You need to use the config file that's specifically for ResNest.

**Comment by kailas711** [comment]
So what i tryin to do is to create custom models.
There are lots of backbones , necks , detectors available in mmdetection in the path "mmdet/models/" as shown below.

![image](https://github.com/open-mmlab/mmdetection/assets/89206677/85571265-7bfd-41ac-9e16-75225748374d)


i want to try different combinations of these. For an object detection models there are a lot of components like **backbone,neck,bounding box regressor,loss etc** are these are available in mmdetection  "mmdet/models/", with that i want to experiment on creating custom models and train them for my research.

**Comment by kailas711** [comment]
as far as i understood if we give the configuration file properly specifying the model components mmdetection will construct them  . Is it possible also i don't want backbones to be pretrained so... is it possible ??

i though once it get the rtmdet working i can use it's config file , becoz idk how to create a new config file , even though i did i don't think it would fit in and work properly

**Comment by kailas711** [comment]
> ```python
> mmengine
> ```
